openni/plugin.cpp

```diff
@@ -0,0 +1,159 @@
+#include "../common/data_format.hpp"
+#include "../common/phonebook.hpp"
+#include "../common/pose_prediction.hpp"
+#include "../common/relative_clock.hpp"
+#include "../common/switchboard.hpp"
+#include "../common/threadloop.hpp"
+
+#include <iomanip>
+#include <opencv/cv.hpp>
+#include <opencv2/core/core.hpp>
+#include <opencv2/highgui/highgui.hpp>
+#include <OpenNI.h>
+
+using namespace ILLIXR;
+
+class openni_plugin : public ILLIXR::threadloop {
+public:
+    openni_plugin(std::string name_, phonebook* pb_)
+        : threadloop{name_, pb_}
+        , sb{pb->lookup_impl<switchboard>()}
+        , _m_clock{pb->lookup_impl<RelativeClock>()}
+        , _m_rgb_depth{sb->get_writer<rgb_depth_type>("rgb_depth")} {
+        if (!camera_initialize()) {
+            std::cout << "Initialization failed" << std::endl;
+            exit(0);
+        }
+    }
+
+    virtual ~openni_plugin() override {
+        color.destroy();
+        depth.destroy();
+    }
+
+protected:
+    virtual skip_option _p_should_skip() override {
+        // read cam data
+        color.readFrame(&colorFrame);
+        depth.readFrame(&depthFrame);
+        // get timestamp
+        assert(colorFrame.getTimestamp() != depthFrame.getTimestamp());
+        cam_time = colorFrame.getTimestamp() * 1000;
```

Comment:
Try to avoid doing work in `_p_should_skip` and do it in `_p_one_iteration` instead. I would try to use the C++ clock instead of the sensor timestamp to tell if it is time to take another picture, but there is some chance that won't work. In that case, look for a lightweight way to get the timestamp from the sensor other than taking a whole picture.